docs/concepts/style_guide.md

```diff
@@ -0,0 +1,206 @@
+
+1. **Readability matters**. If you have to choose between two equally effective
+alternatives, pick the more readable one. This is, of course, subjective, but if
+you can read your own code three months from now, that's a great start. Don't be
+clever over readable, unless you have a documented purpose. Use object oriented
+programming when things get complex and
+[DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself).
+2. **Inherit upstream conventions**. Beaker is still ruby, so use the
+[ruby community style guide](https://github.com/bbatsov/ruby-style-guide). When
+not called out here, use the ruby style guide.
+
+## Test Naming
+
+Tests should test what they say they test. Test names, both the name of the test
+file and the value given to the
+[`test_name`](http://www.rubydoc.info/github/puppetlabs/beaker/Beaker/DSL/Structure#test_name-instance_method)
+function should provide an accurate indication about the purpose of the test.
+
+The `test_name` function should be the first line in the Beaker test file.
+
+**Good:**
+```ruby
+# head -1 puppet/acceptance/tests/resource/file/should_default_mode.rb
+test_name "file resource: set default modes" do
+```
+
+## Structure Methods Should Use Explicit Blocks
+
+These methods aid in self-documenting your tests, including indention in the
+logs. If you don't use explicit blocks, beaker does not know how to properly
+indent your test's output.
+
+The most common
+[structure methods](http://www.rubydoc.info/github/puppetlabs/beaker/Beaker/DSL/Structure)
+are
+[`#test_name`](http://www.rubydoc.info/github/puppetlabs/beaker/Beaker/DSL/Structure#test_name-instance_method),
+[`#step`](http://www.rubydoc.info/github/puppetlabs/beaker/Beaker/DSL/Structure#step-instance_method),
+and
+[`#teardown`](http://www.rubydoc.info/github/puppetlabs/beaker/Beaker/DSL/Structure#teardown-instance_method).
+
+**Good:**
+```ruby
+step 'do this thang' do
+  on(host, "echo 'do this thang'")
+end
+```
+**Bad:**
+```ruby
+step 'do this thang'
+on(host, "echo 'do this thang'")
+```
+
+## Teardowns
+
+- Return the state of the system to the way it was prior to test execution
+- Put the teardown as early in the test as possible
+
+Teardowns must be used to return the system to the state it was in prior to the
+execution or attempted execution of the test. Beaker will gather all teardowns
+encountered throughout the execution path of the test. These teardowns will all
+be executed when the test exits, even if the test exits early due to a failure
+or error.
+
+### Place Teardowns Early
+
+Teardowns can be placed anywhere in the test file or its helpers. The preferred
+style is to have a teardown step near the beginning of the test file to show the
+reader that the system state will be restored.
+
+**Good:**
+```ruby
+test_name 'The source attribute' do
+
+  target_file_on_nix = '/tmp/source_attr_test'
+  teardown do
+    on(hosts[0], "rm #{target_file_on_nix}", :accept_all_exit_codes => true)
```

Comment:
Using an indexed array reference for host does not seem like a great practice to me. If this is the example we want to use, it would seem safer to wrap it in a conditional test for platform type.

Suggestion
```
test_name 'The source attribute' do

  target_file_on_nix = '/tmp/source_attr_test'
  teardown do
    hosts.each do |host|
      on(host, "rm #{target_file_on_nix}", :accept_all_exit_codes => true) unless host['platform'] =~ /^win/
    end
  end
  ...
end
```